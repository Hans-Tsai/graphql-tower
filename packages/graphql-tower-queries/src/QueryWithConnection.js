import { GraphQLInt, GraphQLString, GraphQLList, GraphQLObjectType } from 'graphql';
import assign from 'lodash/assign';
import Query from './Query';

export default class QueryWithConnection extends Query {
  constructor(...args) {
    super(...args);

    this._.args = {
      first: { type: GraphQLInt },
      offset: { type: GraphQLInt },
      after: { type: GraphQLString },
    };

    Object.defineProperty(this, 'args', {
      enumerable: true,
      set: (value) => { assign(this._.args, value); },
      get: () => this._.args,
    });

    Object.defineProperty(this, 'type', {
      enumerable: true,
      set: (type) => {
        this._.node = type;
        this._.type = new GraphQLObjectType({
          name: `${this.name}Connection`,
          fields: {
            totalCount: { type: GraphQLInt },
            nodes: {
              type: new GraphQLList(type),
              resolve: payload => payload,
            },
          },
        });
      },
      get: () => this._.type,
    });
  }
}
