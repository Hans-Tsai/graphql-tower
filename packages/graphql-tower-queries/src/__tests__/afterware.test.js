import _ from 'lodash';
import { GraphQLInt } from 'graphql';
import QueryWithConnection from '../QueryWithConnection';
import { pagination, collections } from '../';

describe('afterware', () => {
  it('pagination', async () => {
    const resolve = jest.fn(() => _.range(1, 2000));
    const afterware = jest.fn((payload, args, context, info, results) => results);

    const QueryPagination = class extends QueryWithConnection {
      type = GraphQLInt;
      resolve = resolve;
      afterware = [pagination, afterware];
    };
    const query = new QueryPagination();
    expect(query).toMatchSnapshot();
    expect(query.type.toString()).toBe('QueryPaginationConnection');

    expect(await query.resolve({}, {})).toEqual(_.range(1, 1001));
    expect(afterware).toHaveBeenCalledTimes(1);

    expect(await query.resolve({}, { first: 100 })).toEqual(_.range(1, 101));
    expect(await query.resolve({}, { offset: 100 })).toEqual(_.range(101, 1101));
    expect(await query.resolve({}, { offset: 100, first: 100 })).toEqual(_.range(101, 201));

    resolve.mockImplementation(() => {
      const results = _.range(1, 2000);
      results.totalCount = 3000;
      results.offset = 50;
      return results;
    });
    expect(await query.resolve({}, { offset: 100, first: 100 }))
      .toEqual(_.assign(_.range(51, 151), { offset: 50, totalCount: 3000 }));

    resolve.mockImplementation(() => _.range(1, 2000).map(id => ({ id: `${id}` })));
    expect(await query.resolve({}, { after: '10' })).toEqual(_.range(10, 1010).map(id => ({ id: `${id}` })));
    expect(await query.resolve({}, { after: '120', first: 100 })).toEqual(_.range(120, 220).map(id => ({ id: `${id}` })));

    resolve.mockImplementation(() => _.range(1, 2000).map(cursor => ({ cursor: `${cursor}` })));
    expect(await query.resolve({}, { after: '120', first: 100 })).toEqual(_.range(120, 220).map(cursor => ({ cursor: `${cursor}` })));
    expect(await query.resolve({}, { after: '3000', first: 200 })).toEqual(_.range(1, 201).map(cursor => ({ cursor: `${cursor}` })));
  });

  it('collections', async () => {
    const fetchPage = jest.fn(() => _.range(1, 100));

    const QueryCollections = class extends QueryWithConnection {
      type = GraphQLInt;
      resolve = async () => ({ fetchPage });
      afterware = [collections];
    };
    const query = new QueryCollections();

    expect(await query.resolve({}, { offset: 100, first: 100 })).toEqual(_.range(1, 100));
    expect(fetchPage).toHaveBeenCalledWith({ offset: 100, first: 100 });
  });
});
