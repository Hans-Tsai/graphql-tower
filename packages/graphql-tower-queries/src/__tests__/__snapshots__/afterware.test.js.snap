// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`afterware pagination 1`] = `
QueryPagination {
  "_": Object {
    "afterware": Array [
      [Function],
      [Function],
    ],
    "args": Object {
      "after": Object {
        "type": "String",
      },
      "first": Object {
        "type": "Int",
      },
      "offset": Object {
        "type": "Int",
      },
    },
    "middleware": Array [],
    "name": undefined,
    "node": "Int",
    "parentArgs": undefined,
    "resolve": [Function],
    "subscribe": [Function],
    "type": "QueryPaginationConnection",
  },
  "args": Object {
    "after": Object {
      "type": "String",
    },
    "first": Object {
      "type": "Int",
    },
    "offset": Object {
      "type": "Int",
    },
  },
  "resolve": [Function],
  "type": "QueryPaginationConnection",
}
`;
